### ISSUE TITLE
Make cluster add-ons work when upgrading the master

### ISSUE THREAD

**Issue by roberthbailey** [GoogleCloudPlatform/kubernetes#8107]
Define who owns cluster add-ons (the actor updating the master or the cluster admin, assuming they are different) and what should happen to them when the master is upgraded:

Should we replace the DNS pods with new ones? Leave the old ones running and document for users how to update them? 

We certainly don't want 2 sets of them running (or to ignore an error trying to start a second set if our intent is to update them). 

Related work in #3579. 

/cc @satnam6502

**Comment by davidopp** [comment]
@satnam6502 would you be interested in working on this?

**Comment by satnam6502** [comment]
@davidopp Yes.

**Comment by davidopp** [comment]
@zmerlynn do you have cycles for this?

**Comment by roberthbailey** [comment]
@zmerlynn told me yesterday that he doesn't currently have cycles. 

@satnam6502 once you investigate please loop @zmerlynn in on the proposed solution since he has thought about this a bit.

**Comment by zmerlynn** [comment]
And feel free to loop me in on what's going to happen when the master is actually upgraded anyways. Currently, the master bits are going to be upgraded pretty forcibly, leaving the Salted templates changed, but the `kube-addons` script isn't going to do the right thing with that.

**Comment by satnam6502** [comment]
**Proposal 1**

As a first very basic step:
* Make it the responsibility of the cluster admin to keep track of any changes made to cluster add-ons relative to their default values. For example, a change to how much memory Kibana is allocated. It is then the cluster-admin's responsibility to re-apply that change after a cluster add-on component is updated.
* Provide command line option to the cluster upgrade command which specify which add-ons to update. For example `--update-all-add-ons` or `--update-add-on elasticsearch-logging`. Add-ons that are not mentioned (except for the all case) are then the responsibility of the cluster admin to update.

**Comment by satnam6502** [comment]
Update:
* I am attempting to work out how to upgrade just the DNS add-on for now. @thockin 
* I will attempt to do this with kubectl rollingupdate.
* During kube-up.sh de-salted YAML config file for the DNS pod only exists at a temporary file somewhere which is cat-ed into `kubectl create` as a part of a loop that processes a bunch of specs. We will probably want to re-factor the kube-up.sh process to identify replication controller pods that are viable for a rolling update. I worry about adding further complexity / bash / salt to an already fragile component. @zmerlynn 
* We could consider an alternative where we at some other step try and perform the upgrade e.g. a step that involves a `kubectl get` for the DNS pod to get its current configuration (and stash it away somewhere); then attempt a rolling update; and if that fails revert to the previous version. This could be a separate Go program we run at the end of the kube-push.sh process, or a utility manually run after the regular kube-push.sh, or we could imagine some "upgrade pod" that is running with the task of noticing when an upgrade needs to be perform to a cluster-add-on and then attempting it.

**Comment by satnam6502** [comment]
**Update.**

My attempts to perform a rolling update of `kube-dsn` are not working out. Here is the result of one experiment. I created a fresh cluster, then from the master I copied `/etc/kubernetes/addons/dns/skydns-rc.yaml` and I changed the label from `kube-dns` to `kube-dns-v0` but otherwise made no changes.
```
kind: ReplicationController
metadata:
  name: kube-dns-v0
  namespace: default
  labels:
    k8s-app: kube-dns-v0
    kubernetes.io/cluster-service: "true"
spec:
  replicas: 1
  selector:
    k8s-app: kube-dns-v0
  template:
    metadata:
      labels:
        k8s-app: kube-dns-v0
```
I then did a rolling update:
```
$ kubectl rolling-update kube-dns -f skydns-rc-v0.yaml 
Creating kube-dns-v0
At beginning of loop: kube-dns replicas: 0, kube-dns-v0 replicas: 1
Updating kube-dns replicas: 0, kube-dns-v0 replicas: 1
At end of loop: kube-dns replicas: 0, kube-dns-v0 replicas: 1
Update succeeded. Deleting kube-dns
kube-dns-v0
```
But this does not seem to have worked:
```
$ kubectl get pods
POD                                            IP           CONTAINER(S)            IMAGE(S)                                             HOST                                     LABELS                                                          STATUS    CREATED      MESSAGE

                                                                          kubernetes-minion-1h14/130.211.115.179   k8s-app=kube-dns-v0,kubernetes.io/cluster-service=true          Pending   4 minutes    
                                                            kube2sky                gcr.io/google_containers/kube2sky:1.4                                                                                                                         Waiting                API error (404): Cannot start container 0388889c9f14c7253d7d6a4e196c779098cbb5de2372066e07d4ebc470660945: [8] System error: exec: "-kubecfg_file=/etc/dns_token/kubeconfig": stat -kubecfg_file=/etc/dns_token/kubeconfig: no such file or directory
...                                                                            
```
which is a complaint about a secret for the dns_token.
And DNS is really busted because after this change the Kibana log viewer complains it can't get data from Elasticsearch (which needs DNS to work).

**Comment by roberthbailey** [comment]
@pmorie since this may be another bug with secrets.

**Comment by pmorie** [comment]
Thanks for heads up, on my list but probably won't have bw til mon

**Comment by davidopp** [comment]
@satnam6502 is going to be out for the next week and a half. @fgrzadkowski said he could find someone to work on this. General approach he is working on is add a go program invoked at the end of kube-push.sh that makes the k8s API calls to update the replication controllers managing the addons. This doesn't work for addons that run as static pods, so something different needs to be done there (modify the manifest on disk or whatever); eventually we would want these pods to run under a per-node controller so nothing is running as a static pod.

**Comment by zmerlynn** [comment]
That approach also doesn't work at all for GKE - it would be better if it were a Go program that integrated at the end of the `kube-addons` process.

**Comment by marekbiskup** [comment]
@pmorie, have you resolved the issue with secrets disappearing after rolling upgrade yet? If not, please create a github separate issue, so that we can track the problem. Thanks!

**Comment by marekbiskup** [comment]
@pmorie I tried to reproduce the problem with rolling update of skydns, in exactly the way @satnam6502 mentioned and I didn't find any problems. I suspect either @satnam6502 had a nasty typo in his yaml, or that version of kube2sky was broken. I believe there is no problem. But if you reproduce the bug, just file another issue.

**Comment by pmorie** [comment]
@marekbiskup @satnam6502 secrets integrity across node reboot should be fixed now

**Comment by marekbiskup** [comment]
The plan is to createc a new binary kube-apiserver.go, deployed on kubernetes-master. The binary will:
  1. Read the directory $1 (/etc/kubernetes/addons) and load api objects to memory
  1. Query the API server to check which objects are available with label "kubernetes.io/cluster-service=true" (services, replication-controllers)
  1. Match the existing api objects and objects from the directory
     1. Use object name (exact, for instance kibana-logging-v1 and kibana-logging-v2 **are not** the same)
  1. If an object exists only in the API, delete it (retry a few times)
  1. If an objects exists only in the dir, create it
  1. If an object exists in both
    1. If they are exactly the same, just ignore them
    1. For replication controllers, perform rolling upgrade   (**in the first version I'll perform delete, (also pods) + create**)
    1. For services, delete the old one and create the new one.

In the first iteration, the program will be invoked from /etc/kubernetes/kube-addons.sh

**Comment by roberthbailey** [comment]
@marekbiskup - Just a nit, but I assume you didn't mean to say that the binary would be called kube-apiserver.go (since that name is already taken). 

It sounds like this is another controller loop. Did you talk to @lavalamp about whether there is an existing place on the master that it could be added (e.g. the controller-manager)?

**Comment by davidopp** [comment]
Plan SGTM, just a few questions
1) how do the JSON files in /etc/kubernetes/addons get created?
2) what is your plan for updating addons that run as static pods (currently just fluentd)
3) what happens if the binary you're creating (I agree with @roberthbailey, you should not call it kube-apiserver.go :-) crashes before the updating is finished?

Have you considered the alternative of just driving this all from a command-line script rather than creating a binary that runs on the master? I thought when we discussed this in person we were leaning towards a script the user would run rather than doing something on the server. It seems simpler.

**Comment by zmerlynn** [comment]
> 1) how do the JSON files in /etc/kubernetes/addons get created?

Presumably Salt, as today. We don't need to change that yet - one of the issues this system avoids (and @marekbiskup is relaying a design I worked on that actually solves this issue more or less correctly) is that it doesn't actually need to solve the global configuration issue, it just need reconcile add-on manifest created by Salt, which Salt already knows how to do.

> 2) what is your plan for updating addons that run as static pods (currently just fluentd)

This is already handled for rolling node recreates / master teardown style upgrades, so .. done. (The manifest just changes.) For binary in-place upgrade, we might need to file an issue for later. (This work may not need to be done for now. It depends on outcome of #8789.)

> 3) what happens if the binary you're creating (I agree with @roberthbailey, you should not call it kube-apiserver.go :-) crashes before the updating is finished?

I was actually hoping this was a controller loop somewhere else that was monitored....

**Comment by roberthbailey** [comment]
> I was actually hoping this was a controller loop somewhere else that was monitored....

Me too. A controller loop in a new pod managed by the kubelet or as part of the existing controller-manager pod means that we could count on the kubelet making sure it was always reconciling the state of the cluster addons.

**Comment by fgrzadkowski** [comment]
@marekbiskup is OOO today, but my understanding is that his plan was to indeed run it as a pod. I think that the step he described (update current script) is just an intermediate step towards full solution.

**Comment by davidopp** [comment]
Thanks for the answers. I still think it would be a good idea to prototype this as a shell script that just invokes kubectl. Besides working out bugs in the overall idea, it would ensure we have something in time for the 1.0 deadline (I'm a little concerned that there might not be enough time to do the full-blown server-based version).

**Comment by zmerlynn** [comment]
A shell script driven by the user? That's mostly a non-starter for GKE for the same reason that `kube-addon` was moved to the server in the first place - we don't want to have to imitate the heavy client logic.

**Comment by davidopp** [comment]
@marekbiskup Can you summarize your progress on this? This is a pretty big project given the time constraints; is it possible to parallelize the work?

**Comment by marekbiskup** [comment]
I decided to follow @davidopp's suggestion to have a bash script instead of a go binary. The solution will be exactly what @zmerlynn described on a meeting some 2 weeks ago. This is just a tiny change to what we have now. 

Here is a description that I hope will be more readable

# What we have:
 * addon manifests are deployed (by salt) in /etc/kubernetes/addons/ on the master (by kube-push, or just be creating a new VM in GKE)
 * linux service /etc/init.d/kube-addons (written in bash), started with the OS. It reads yamls and jsons from /etc/kubernetes/addons/ (once, at start) and tries to create them on the API server

### problems
if addon manifests are updated (by GKE, or just kube-pushed), kube-addons (if restarted only) will try create the addons again and
1. old objects will not be deleted
2. new objects may not be created if the name didn't change

### proposed solution
kube-addons will call periodically addon-update.sh (every 10 minutes) and do more or less what I stated in my comment above (why doesn't github number comments!). But slightly changed:

1. (The script will handle only replication controllers and services.)
1. Read the directory /etc/kubernetes/addons and find the manifests (only yaml, for start)
1. Query the API server to check which objects are available with label "kubernetes.io/cluster-service=true" (services, replication-controllers)
1. Match the existing api objects against manifests
  1. Use object base name (to the last dash -), and version (after the last dash) e.g. kibana-logging-v1 has kibana-logging as the base name and v1 as version
    1. In the first version (as I'm writing it in bash), I'll take the *file name* of the manifest as the object name
  1. If an object exists only in the API, delete it (retry a few times)
  2. If an objects exists only in a manifest, create it
  3. If an object exists in both (i.e. basename are the same)
    1. If the versions are the same, just ignore them (in the future: also compare other data of the objectsm e.g. to find new labels)
    1. If versions differ, stop the one on the API server and create the new from the manifest (in the future: perform rolling upgrade of replication controllers)

In addition, I'll take care of the following
1. Monitor  /etc/init.d/kube-addons using monit
1. Make sure 'service kube-addons restart' is performed after kube-push (Important if we push new version of /etc/init.d/kube-addons
1. Rename the manifests so that file names match object names (and make the names consistent)

### future
1. Make the script more robust (limitations will be written in the code) - I suggest rewriting it in go.
2. Move the service to a pod and run it as a master-addon
3. perform rolling-update of pods instead of stop-create

# Things that may be questioned (suggestions welcomed)
1. Why run it in a loop, isn't it enough to run it once (after restart or after kube-push)

**Comment by satnam6502** [comment]
Do you need #8469 to be merged (or something similar) before this will work?
This change adds a version suffix to the name of the monitoring pod and adds labels to support a rolling update.

**Comment by marekbiskup** [comment]
@satnam6502, 
Yes. 
I need even more - I need file names to be the same as metadata.name (unless somebody tells me how to extract metadata.name from a yaml file in bash without installing additional tools)

Are you going to merge it, or shall I take your patch and work on it? (It will save me a lot of time anyway)

**Comment by satnam6502** [comment]
I think someone should merge it (and see what happens -- keeping a close eye on e2e).
I can't merge it since it is my PR. @erictune @vishh have been commenting on it and it has a kind-of LGTM.

**Comment by marekbiskup** [comment]
Ok, I cherry-picked your commit and the merge is not blocking me.
Thanks for pointing me to this change!

**Comment by marekbiskup** [comment]
One more thing - the proposed solution works for replication controllers, but not for services (we don't want service name change). For now services update will not work (we should decide if want it in v1.0)

**Comment by zmerlynn** [comment]
If you're going down this route and you want to extract yaml in bash *on
the cluster*, you can safely use inline python in your bash and `import
yaml`. We've gotten into trouble with that package on the client side
because it's not a default package everywhere, but on all Salt based
clusters it's required by Salt, so for this solution in particular, you
know it's available.

**Comment by marekbiskup** [comment]
@zmerlynn, it works perfect. So I'll ignore file names - just metadata.name will be taken into account.
I'll update the previous comment

**Comment by marekbiskup** [comment]
Yet another thing. I'm configuring salt to delete all objects under /etc/kubernetes/addons before deploying new manifests under this directory. This means that all files that were put here as addons, manually, will be deleted at kube-push. And later removed from the API server. I think this is what we need.

**Comment by zmerlynn** [comment]
@marekbiskup: Thank you! That was something I spent a little headbanging time trying to figure out how to do, but it wasn't intuitive how to get Salt to clean a whole directory. Did you find an easy pattern for it?

**Comment by marekbiskup** [comment]
Stackoverflow hepled me to figure it out:
```
--- a/cluster/saltbase/salt/kube-addons/init.sls
+++ b/cluster/saltbase/salt/kube-addons/init.sls
@@ -1,3 +1,16 @@
+addon-dir-delete:
+  file.absent:
+    - name: /etc/kubernetes/addons
+
+addon-dir-create:
+  file.directory:
+    - name: /etc/kubernetes/addons
+    - user: root
+    - group: root
+    - mode: 0755
+    - require:
+        - file: addon-dir-delete
+
```

**Comment by roberthbailey** [comment]
My only concern with bash vs. a Go program is that our bash scripts are really, really terrible at error checking and handling error flows. Using `set -o errexit` is great but causes the script to bail out in the middle (and doesn't apply to subshells created using `$()`).

**Comment by zmerlynn** [comment]
I'm fine with a bash script this week if you're actually intending a Go
program soon. I honestly thought we had enough time to land one given the
modest requirements here, especially if we integrated it into an existing
controller, and given that not upgrading these components is really a bug
and not a feature...

However, I'm also fine with a bash script for 1.0 and a Go program for 1.1,
too.

**Comment by davidopp** [comment]
We're supposed to be code complete for new features by Friday and the Poland guys have a public holiday Thu-Fri so I think we should stick with the bash script for 1.0.  (Though I totally agree that a real go binary would be better and we should do it for next version. Hopefully @marekbiskup can start working on that once this goes in).

**Comment by davidopp** [comment]
(#9128 is out for review)

**Comment by marekbiskup** [comment]
I'll be happy to rewrite it in go, then put it into a controller to be run as a pod. Finally, to perform rolling upgrade. For 1.1 of course. For 1.0, I'll implement e2e tests, so that we can track regression. I'll totally skip unittests for bash.

**Comment by marekbiskup** [comment]
I haven't found any way to tell salt to restart the linux service kube-addons. Anyone has any idea?
I'll file another bug for this if my PR is merged.

**Comment by zmerlynn** [comment]
It needs a reason to restart it, like managed files. It would be obnoxious,
but maybe the whole template list if you can't give it the directory?

**Comment by marekbiskup** [comment]
in #9128 me and @zmerlynn were discussing whether version strings should be included in rc names.

A WIP version with version strings is here:
https://github.com/marekbiskup/kubernetes/compare/addonUpgrade...marekbiskup:addonUpgradeVersionLabel?expand=1

(this is a diff from the tip of #9128)
cc @bgrant0607

**Comment by alex-mohr** [comment]
Poland has had a national holiday the past couple days; ETA landing M/T next week.  Marking P0 as a hard RC blocker.

**Comment by marekbiskup** [comment]
@bgrant0607 clarified (in a personal email) that version string ```should be included``` in RC names

**Comment by marekbiskup** [comment]
Merged in #9128:

 * linux service (kube-addon) for monitoring add-on dir on the master
   * if files change, API objects are updated accordingly (stop/start, no rolling upgrade)
 * add-on dir is polled every 10 minutes
 * kube-push restarts the service, so it checks the add-on dir without the (up to) 10-minute delay
 * the log of the kube-addon service is rotated by logrotate
 * only manual tests were performed
 * services and RCs are not updated unless their name changes (in case of RC it is enough because the version is included in the name; for services, this is a missing functionality

Folow-ups:
 * #9394 - automated tests (for 1.0)
 * #9463 - rolling update for RC and update of services (after 1.0 - maybe 1.1)
